File: src/0768/S0768_v1.py
class Solution:
    def maxChunksToSorted(self, arr: List[int]) -> int:
        rec = [None] * (len(arr)+1)
        rec[0] = 0
        cur, p = arr[0], 0
        for i in arr:
            while i < rec[p]:
                p -= 1
            p += 1
            if i > cur:
                cur = i;
            rec[p] = cur
        return p

File: src/0768/S0768_v1.cpp
class Solution {
public:
    int maxChunksToSorted(vector<int>& arr) {
        int rec[arr.size()+1];
        rec[0] = 0;
        int cur = arr[0];
        int p = 0;
        for (int i = 0; i < arr.size(); i++) {
            while (arr[i] < rec[p]) p--;
            if (arr[i] > cur) cur = arr[i];
            rec[++p] = cur;
        }
        return p;
    }
};
